{"metadata": {"kernelspec": {"display_name": "Python 3 (ipykernel)", "language": "python", "name": "python3"}, "language_info": {"codemirror_mode": {"name": "ipython", "version": 3}, "file_extension": ".py", "mimetype": "text/x-python", "name": "python", "nbconvert_exporter": "python", "pygments_lexer": "ipython3", "version": "3.9.16"}}, "nbformat_minor": 5, "nbformat": 4, "cells": [{"cell_type": "markdown", "source": "# Extract SCN9A Genotype Data From UK BioBank\nAuthor: Graeme Newton - graeme.newton@bristol.ac.uk / gwtnewton@outlook.com \\\nDate: 10/05/2024\n\nThis code should be run via a jupytrlab notebook on DNAnexus. \\\nIf errors are encountered while running this script, it is likely due to insufficient RAM for annotation. For VEP, lower the --fork and or --buffer_size to fix this, or use a more powerful instance. \\\n\nRecommended instance: 2 x mem3_ssd2_v2_x4 \\\nApproximate runtime: 60 minutes \\\nApproximate cost: \u00a30.20", "metadata": {}, "id": "eebcec64-0b0c-45fa-b18b-c29db3006d47"}, {"cell_type": "code", "source": "%%bash\n# Function to download PLINK2\ndownload_plink2() {\n  wget https://s3.amazonaws.com/plink2-assets/plink2_linux_avx2_20240504.zip \\\n    && unzip plink2_linux_avx2_20240504.zip -d plink2_tmp \\\n    && mv plink2_tmp/plink2 \"${INSTALL_DIR}/bin\" \\\n    && rm -rf plink2_linux_avx2_20240504.zip plink2_tmp\n}\n\n# Function to download data using dx\ndownload_data() {\n  dx download \"Bulk/Exome sequences/Population level exome OQFE variants, PLINK format - final release/ukb23158_c2_b0_v1.*\"\n}\n\n# Send all outputs during installation to the log file\nexec > installation_log.txt 2>&1\n\n# Run PLINK2 download function in background\ndownload_plink2 &\n\n# Run data download function in background\ndownload_data &\n\n# Install sudo annd other in foreground due to installation order dependencies\nwget https://www.sudo.ws/dist/sudo-1.9.15p5.tar.gz \\\n  && tar xf sudo-1.9.15p5.tar.gz \\\n  && cd sudo-1.9.15p5 \\\n  && ./configure && make && make install \\\n  && cd .. && rm -f sudo-1.9.15p5.tar.gz && rm -rf sudo-1.9.15p5\n\n# Install build-essential package\nsudo apt update\nsudo apt install -y build-essential libmysqlclient-dev\n\n# Install Perl modules using cpanminus\ncurl -L https://cpanmin.us | sudo perl - App::cpanminus\nsudo cpanm Archive::Zip DBI DBD::mysql Bio::Root::Version Bio::DB::HTS\n\n# Clone and setup VEP\ngit clone https://github.com/Ensembl/ensembl-vep.git\ncd ensembl-vep\nperl INSTALL.pl --AUTO acfp --ASSEMBLY GRCh38 --PLUGINS all --SPECIES homo_sapiens\n\n# Download Homo sapiens cache data for VEP\ncurl -O https://ftp.ensembl.org/pub/release-111/variation/indexed_vep_cache/homo_sapiens_vep_111_GRCh38.tar.gz\ntar xzf homo_sapiens_vep_111_GRCh38.tar.gz\nrm homo_sapiens_vep_111_GRCh38.tar.gz\ncd ..", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "6584fd35-8b63-4bb6-87b9-67b3d1aa3d47"}, {"cell_type": "code", "source": "%%bash\nplink2 --bfile ukb23158_c2_b0_v1 --chr 2 --from-bp 166185000 --to-bp 166386000 --no-categorical --geno-counts --sample-counts --recode vcf --out SCN9A_gene", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "e39cbf35-cd93-4256-a8a7-89c23a4e5f92"}, {"cell_type": "code", "source": "%%bash\n# Annotate the SCN9A vcf file\n./ensembl-vep/vep -i ./SCN9A_gene.vcf --fork 4 --cache --dir_cache ./ensembl-vep --species homo_sapiens --pick --everything --canonical --vcf --buffer_size 100 --force_overwrite -o SCN9A_gene_anno.vcf", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "8d33f4b1-ec2d-463e-9548-88309b70ab15"}, {"cell_type": "code", "source": "%%bash\n# Compress the genotype files\ngzip SCN9A_gene*.vcf\n\n# Upload alloutput files to your DNAnexus project\ndx upload SCN9A_gene*", "metadata": {"trusted": true, "tags": []}, "execution_count": null, "outputs": [], "id": "e4d8e756-331d-45bb-bdda-9e5fbf1e6981"}]}